adcc/workflow.py

```diff
@@ -495,3 +505,85 @@ def setup_solver_printing(solmethod_name, matrix, kind, default_print,
         def inner_callback(state, identifier):
             default_print(state, identifier, output)
         return inner_callback
+
+
+def setup_solvent(matrix, solvent_scheme):
+    """
+    Setup solvent matrix terms and/or energy corrections.
+    Internal function called from run_adc.
+    """
+    # TODO: move someplace meaningful & reasonable...
```

Comment:
Yes let's discuss ...

Reply:
My idea was a table in the docs with the long explanations and references to papers; plus a short description in the function documentation that one can view via `adcc.workflow.setup_solvent?`

What do you think?

Reply:
Sounds perfect. Keywords should be short and descriptive. Perhaps they should actually be a class / named tuple (this way you can take optional "arguments" in case that might be necessary for more involved models, like how many times you iterate a response or which order of perturbation theory or sth like that).

Reply:
What exactly should be a class or namedtuple? The collection of all keywords/options or one instance of the namedtuple for each scheme?
I think for now one could also stick with the scheme "names" and if we need more advanced settings in the future, these can be enabled via `**kwargs`, which is the usual way in `adcc` I guess.

Reply:
Nope what I meant was instead of `solvent_model="ptss"` you do `solvent_model=ptss` which is an actually `ptss` class or named tuple. In that way you can later do `solvent_model=ptss(order=5)` in case we ever implemented 5-th order corrections.

Reply:
For me this feels too complicated for now... I think what you want could be introduced without breaking changes if we ever need this in the future, i.e., allow for either `solvent_model=["ptss", "ptlr"]` which chooses pre-defined configurations or provides full flexibility via `solvent_model=[ptss(order=100), ptlr(order=10)]`.

I don't know... I have to think about it. Probably one would not want to write more `import` statements in the Python script to run `adcc` just to be able to use the corrections.

Reply:
Yes that's true. As long as there is a reasonable way to later include this I'm ok. I just want to avoid that we have 10000 keywords in run_adc on the one hand and 1000 possible combinations of solvent keywords on the other, that's why I threw that comment in. I agree that for now supporting a few strings is ok. But a list of strings already feels like something more involved, where a class with sensible arguments could be a more pythonic alternative, which we should consider for a sec.

Reply:
Difficult design choice... I have no clue which alternative is best in the long run.
If we go for a class/namedtuple to configure everything, how would that look?
This is the same ol' "struct of arrays" or "array of structs" problem... 

One scenario:
```Python
config = SolventConfig(ptss=True, ptlr=True)
adcc.adc2(scfres, n_singlets=5, solvent_config=config)
```

Second scenario:
```Python
config = [
    SolventConfig("ptss"), SolventConfig("ptlr")
]
```
or whatever.

The only thing I don't like about the current solution (string or list of strings) is that it's pretty
unclear what you are allowed to do. I mean the code will raise Exceptions if you select an invalid combination, but
it's a bit weird.

I mean you still have full flexibility and can compose anything you like via the new classes `EnergyCorrection` and
`AdcExtraTerm`.

Reply:
I would do neither. I'd do
```solvent_config=[ptss, ptlr]```
which could also accept a list of strings ... but then we are sort of back where we started. I don't know. I'm just ultimately not happy about strings because it gives infinite possibilities (also typos etc.) and it does not suggest what is valid and what not at all. Flags (like `(ptss=True, ptlr=True)` are typically much more meaningful and better for tab completion.

Reply:
Okay probably `SolventConfig(ptss=True, ptlr=True)` or so is the easiest and cleanest way to handle it.
The class could also take care of basic validation, i.e., which combinations are allowed and which not in general, e.g.,
`SolventConfig(ptss=True, ptlr=True, postscf=True)` should throw an error.
`SolventConfig(ptss=True, hf=True)` as well. So the class attributes don't suggest which *combinations* are valid,
but at least which keywords...


If we decide to go this way, we should disable the list of strings stuff.

Reply:
Yes that's true. Otherwise it is too much. Also we should have a sane default and `SolventConfig` is not the best name ...

Reply:
The "sane default" is the only thing that is still bugging me... maybe I'll have better ideas tomorrow 🙄

Reply:
I have to be frank... I don't like either solution for this problem. Maybe the solvent stuff needs to become "advanced" use, i.e., you need to do everything manually...

The reason why I'm saying this is:

Let's say `cfg = SolventConfig()` is by default `SolventConfig(hf=True)`, i.e., coupling only through "solvated orbitals" nothing further. Then `cfg = SolventConfig(ptss=True)` would be invalid (HF + ptSS is mutually exclusive), and one would need to specify `cfg = SolventConfig(hf=False, ptss=True)`, which is a lot of code to type for
what it achieves.
`cfg = SolventConfig(hf=False)` would be invalid later on *if* you have an SCF result w/ solvent attached. And so on and so forth... the whole logic of this in itself is not user-friendly at all.

Reply:
Perhaps the `solvent` argument takes either `None` (auto-decided), `False` (HF only) or `Correction(ptss=True, ptlr=True)` or something along those lines?

Reply:
I'm not sure... different behavior for `None` and `False` does not seem a good choice, and adding a class `HfOnly` that basically does not do anything except "storing" the information just seems too much.

Maybe discuss via Skype?

Reply:
Ok, but not now. At 13?